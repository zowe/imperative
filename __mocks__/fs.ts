/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

const fs = jest.genMockFromModule("fs") as any;
const origReadFileSync = jest.requireActual("fs").readFileSync;
const mockReadFileSync = fs.readFileSync;

function readFileSync(filePath: string, encoding?: string) {
    // Don't mock if yargs is trying to load a locale json file
    if (filePath.match(/node_modules.yargs/)) {
        return origReadFileSync(filePath, encoding);
    }
    return mockReadFileSync(filePath, encoding);
}

fs.readFileSync = readFileSync;

module.exports = fs;
