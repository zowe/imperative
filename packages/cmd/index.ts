/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/doc/profiles/definition/ICommandProfileSchema";
export * from "./src/doc/profiles/definition/ICommandProfileProperty";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";
export * from "./src/doc/handler/ICommandHandler";
export * from "./src/doc/handler/ICommandHandlerResponseChecker";
export * from "./src/doc/handler/ICommandHandlerResponseValidator";
export * from "./src/doc/handler/IHandlerParameters";
export * from "./src/doc/handler/IChainedHandlerEntry";
export * from "./src/doc/handler/IChainedHandlerArgumentMapping";

export * from "./src/doc/option/ICommandOptionAllowableValues";
export * from "./src/doc/option/ICommandOptionDefinition";
export * from "./src/doc/option/ICommandOptionValueImplications";
export * from "./src/doc/option/ICommandPositionalDefinition";

export * from "./src/doc/response/response/ICommandResponse";
export * from "./src/doc/response/parms/ICommandResponseParms";
export * from "./src/doc/response/response/ICommandValidatorError";
export * from "./src/doc/response/response/ICommandValidatorResponse";

export * from "./src/doc/ICommandDefinition";
export * from "./src/doc/ICommandDefinitionPassOn";
export * from "./src/doc/ICommandExampleDefinition";
export * from "./src/doc/IPartialCommandDefinition";

export * from "./src/doc/args/ICommandArguments";

export * from "./src/handlers/FailedCommandHandler";

export * from "./src/help/abstract/AbstractHelpGenerator";
export * from "./src/help/abstract/AbstractHelpGeneratorFactory";

export * from "./src/help/doc/IHelpGeneratorFactory";
export * from "./src/help/doc/IHelpGeneratorFactoryParms";
export * from "./src/help/doc/IHelpGeneratorParms";
export * from "./src/help/DefaultHelpGenerator";
export * from "./src/help/HelpConstants";
export * from "./src/help/HelpGeneratorFactory";
export * from "./src/help/doc/IHelpGenerator";
export * from "./src/help/WebHelpGenerator";
export * from "./src/help/WebHelpManager";

export * from "./src/doc/profiles/definition/ICommandProfile";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";

export * from "./src/profiles/CliProfileManager";

export * from "./src/syntax/SyntaxValidator";

export * from "./src/utils/CommandUtils";
export * from "./src/utils/SharedOptions";

export * from "./src/yargs/doc/IYargsParms";
export * from "./src/yargs/doc/IYargsResponse";

export * from "./src/yargs/AbstractCommandYargs";
export * from "./src/yargs/CommandYargs";
export * from "./src/yargs/GroupCommandYargs";
export * from "./src/yargs/YargsConfigurer";
export * from "./src/yargs/YargsDefiner";

export * from "./src/CommandPreparer";
export * from "./src/CommandProcessor";
export * from "./src/response/CommandResponse";

export * from "./src/profiles/CommandProfiles";

export * from "./src/response/CommandResponse";
export * from "./src/response/HandlerResponse";

export * from "./src/doc/response/api/handler/IHandlerResponseApi";
export * from "./src/doc/response/api/handler/IHandlerProgressApi";
export * from "./src/doc/response/api/handler/IHandlerResponseConsoleApi";
export * from "./src/doc/response/api/handler/IHandlerResponseDataApi";
export * from "./src/doc/response/api/handler/IHandlerFormatOutputApi";

export * from "./src/doc/response/response/ICommandOutputFormat";
