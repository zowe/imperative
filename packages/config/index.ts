/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/Config";
export * from "./src/ConfigAutoStore";
export * from "./src/ConfigConstants";
export * from "./src/ConfigSchema";
export * from "./src/ConfigBuilder";
export * as ConfigUtils from "./src/ConfigUtils";
export * from "./src/ProfileCredentials";
export * from "./src/ProfileInfo";
export * from "./src/ProfInfoErr";
export * from "./src/doc/IConfig";
export * from "./src/doc/IConfigAutoStoreOpts";
export * from "./src/doc/IConfigBuilderOpts";
export * from "./src/doc/IConfigConvertResult";
export * from "./src/doc/IConfigLayer";
export * from "./src/doc/IConfigOpts";
export * from "./src/doc/IConfigProfile";
export * from "./src/doc/IConfigSchema";
export * from "./src/doc/IConfigSecure";
export * from "./src/doc/IConfigVault";
export * from "./src/doc/IProfArgAttrs";
export * from "./src/doc/IProfAttrs";
export * from "./src/doc/IProfInfoUpdatePropOpts";
export * from "./src/doc/IProfInfoRemoveKnownPropOpts";
export * from "./src/doc/IProfInfoErrParms";
export * from "./src/doc/IProfLoc";
export * from "./src/doc/IProfMergeArgOpts";
export * from "./src/doc/IProfMergedArg";
export * from "./src/doc/IProfInfoProps";
export * from "./src/doc/IProfOpts";
