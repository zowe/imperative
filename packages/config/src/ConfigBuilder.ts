/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { IImperativeConfig } from "../../imperative";
import { Config } from "./Config";
import { IConfig } from "./doc/IConfig";
import { IConfigBuilderOpts } from "./doc/IConfigBuilderOpts";
import { IConfigProfile } from "./doc/IConfigProfile";

export class ConfigBuilder {
    /**
     * Build a new Config object from an Imperative CLI app configuration.
     * @param impConfig The Imperative CLI app configuration.
     * @param opts Options to control aspects of the builder.
     */
    public static async build(impConfig: IImperativeConfig, opts?: IConfigBuilderOpts): Promise<IConfig> {
        opts = opts || {};
        const config: IConfig = Config.empty();

        for (const profile of impConfig.profiles) {
            const properties: { [key: string]: any } = {};
            const secureProps: string[] = [];
            for (const [k, v] of Object.entries(profile.schema.properties)) {
                if (opts.populateProperties && v.includeInTemplate) {
                    if (v.secure) {
                        secureProps.push(k);
                        const propValue = opts.getSecureValue ? await opts.getSecureValue(k, v) : null;
                        if (propValue != null) {
                            // Save this value to be stored securely after profile is built
                            properties[k] = propValue;
                        }
                    } else {
                        if ((v as any).optionDefinition != null) {
                            // Use default value of ICommandOptionDefinition if present
                            properties[k] = (v as any).optionDefinition.defaultValue;
                        }
                        if (properties[k] === undefined) {
                            // Fall back to an empty value
                            properties[k] = this.getDefaultValue(v.type);
                        }
                    }
                }
            }

            // Add the profile to config and set it as default
            lodash.set(config, `profiles.${profile.type}`, {
                type: profile.type,
                properties,
                secure: secureProps
            });

            if (opts.populateProperties) {
                config.defaults[profile.type] = profile.type;
            }
        }

        // Hoist duplicate default properties
        if (config.profiles != null && impConfig.baseProfile != null) {
            this.hoistTemplateProperties(config.profiles, impConfig.baseProfile.type);
        }

        return { ...config, autoStore: true };
    }

    /**
     * Returns empty value that is appropriate for the property type.
     * @param propType The type of profile property
     * @returns Null or empty object
     */
    private static getDefaultValue(propType: string | string[]): any {
        // TODO How to handle profile property with multiple types
        if (Array.isArray(propType)) {
            propType = propType[0];
        }
        switch (propType) {
            case "string":  return "";
            case "number":  return 0;
            case "object":  return {};
            case "array":   return [];
            case "boolean": return false;
            default:        return null;
        }
    }

    /**
     * Moves duplicate default properties across child profiles into a base
     * profile.
     * @param profiles The profiles object of the config JSON
     * @param baseProfileType The type of the base profile
     */
    private static hoistTemplateProperties(profiles: { [key: string]: IConfigProfile }, baseProfileType: string): void {
        // Flatten properties into object that maps property name to list of values
        const flattenedProps: { [key: string]: any[] } = {};
        for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
            for (const [k, v] of Object.entries(childProfile.properties)) {
                flattenedProps[k] = [...(flattenedProps[k] || []), v];
            }
        }
        // List property names defined multiple times with the same value
        const duplicateProps: string[] = [];
        for (const [k, v] of Object.entries(flattenedProps)) {
            if (v.length > 1 && (new Set(v)).size === 1) {
                duplicateProps.push(k);
            }
        }
        // Remove duplicate properties from child profiles and store them in root profile
        for (const propName of duplicateProps) {
            profiles[baseProfileType].properties = {
                [propName]: flattenedProps[propName][0],
                ...profiles[baseProfileType].properties
            };
            for (const childProfile of Object.values(profiles).filter(({ type }) => type !== baseProfileType)) {
                delete childProfile.properties[propName];
            }
        }
    }
}
