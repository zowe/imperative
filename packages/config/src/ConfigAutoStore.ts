/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { ICommandArguments, IHandlerParameters } from "../../cmd";
import { ICommandHandlerRequire } from "../../cmd/src/doc/handler/ICommandHandlerRequire";
import { ICommandProfileAuthConfig } from "../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { IConfigLoadedProfile } from "./doc/IConfigLoadedProfile";
import * as ConfigUtils from "./ConfigUtils";
import { AbstractAuthHandler } from "../../imperative/src/auth/handlers/AbstractAuthHandler";
import { ImperativeConfig } from "../../utilities";
import { ISession } from "../../rest/src/session/doc/ISession";
import { Session } from "../../rest/src/session/Session";
import { AUTH_TYPE_TOKEN } from "../../rest/src/session/SessConstants";
import { Logger } from "../../logger";

/**
 * Class to manage automatic storage of properties in team config.
 */
export class ConfigAutoStore {
    /**
     * Finds the token auth handler class for a team config profile.
     * @param profilePath JSON path of profile
     * @param cmdArguments CLI arguments which may specify a profile
     * @returns Auth handler class or undefined if none was found
     */
    public static findAuthHandlerForProfile(profilePath: string, cmdArguments: ICommandArguments): AbstractAuthHandler | undefined {
        const config = ImperativeConfig.instance.config;
        const profileType = lodash.get(config.properties, `${profilePath}.type`);
        const profile = config.api.profiles.get(profilePath.replace(/profiles\./g, ""));

        if (profile == null || profileType == null) {  // Profile must exist and have type defined
            return;
        } else if (profileType === "base") {
            if (profile.tokenType == null) {  // Base profile must have tokenType defined
                return;
            }
        } else {
            if (profile.basePath == null) {  // Service profiles must have basePath defined
                return;
            } else if (profile.tokenType == null) {  // If tokenType undefined in service profile, fall back to base profile
                const baseProfileName = ConfigUtils.getActiveProfileName("base", cmdArguments);
                return this.findAuthHandlerForProfile(config.api.profiles.expandPath(baseProfileName), cmdArguments);
            }
        }

        const authConfigs: ICommandProfileAuthConfig[] = [];
        ImperativeConfig.instance.loadedConfig.profiles.forEach((profCfg) => {
            if ((profCfg.type === profileType || profCfg.type === "base") && profCfg.authConfig != null) {
                authConfigs.push(...profCfg.authConfig);
            }
        });

        for (const authConfig of authConfigs) {
            const authHandler: ICommandHandlerRequire = require(authConfig.handler);
            const authHandlerClass = new authHandler.default();

            if (authHandlerClass instanceof AbstractAuthHandler) {
                const promptParams = authHandlerClass.getPromptParams()[0];

                if (profile.tokenType === promptParams.defaultTokenType) {
                    return authHandlerClass;  // Auth service must have matching token type
                }
            }
        }
    }

    /**
     * Finds the highest priority layer where a profile is stored.
     * @param loadedProfile
     * @returns User and global properties
     */
    public static getPriorityLayer(loadedProfile: IConfigLoadedProfile): { user: boolean, global: boolean } {
        return {
            user: Object.values(loadedProfile.properties).every(v => v.user),
            global: Object.values(loadedProfile.properties).some(v => v.global)
        };
    }

    /**
     * Stores session config properties into a team config profile.
     * @param params CLI handler parameters object
     * @param sessCfg Session config containing properties to store
     * @param propsToStore Names of properties that should be stored
     */
    public static async storeSessCfgProps(params: IHandlerParameters, sessCfg: { [key: string]: any }, propsToStore: string[]): Promise<void> {
        const config = ImperativeConfig.instance.config;
        // TODO Which autoStore value should take priority if it conflicts between layers
        if (propsToStore.length == 0 || !config?.exists || !config.properties.autoStore) {
            return;
        }

        let profileProps = propsToStore.map(propName => propName === "hostname" ? "host" : propName);
        const profileData = this.findActiveProfile(params, profileProps);
        if (profileData == null) {
            return;
        }
        const [profileType, profileName] = profileData;
        const profilePath = config.api.profiles.expandPath(profileName);

        // Replace user and password with tokenValue if tokenType is defined in config
        if (profileProps.includes("user") && profileProps.includes("password") &&
            await this.fetchTokenForSessCfg(params, sessCfg, profilePath)) {
            profileProps = profileProps.filter(propName => propName !== "user" && propName !== "password");
            profileProps.push("tokenValue");
        }

        const beforeLayer = config.api.layers.get();
        const loadedProfile = config.api.profiles.load(profileName);
        const { user, global } = this.getPriorityLayer(loadedProfile);
        config.api.layers.activate(user, global);

        const baseProfileName = ConfigUtils.getActiveProfileName("base", params.arguments);
        const baseProfileObj = lodash.get(config.properties, config.api.profiles.expandPath(baseProfileName));
        const baseProfileSchema = ImperativeConfig.instance.loadedConfig.baseProfile.schema;
        const profileSchema = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profileType).schema;

        for (const propName of profileProps) {
            let propProfilePath = profilePath;
            // Determine if property should be stored in base profile instead
            if (loadedProfile.properties[propName] == null && !loadedProfile.secure?.includes(propName) &&
                (baseProfileObj.properties[propName] != null || baseProfileObj.secure?.includes(propName) ||
                (propName === "tokenValue" && baseProfileObj.properties.tokenType != null))) {
                propProfilePath = config.api.profiles.expandPath(baseProfileName);
            }
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            const propDefinition = profileSchema.properties[propName] || baseProfileSchema.properties[propName];
            config.set(`${propProfilePath}.properties.${propName}`, sessCfg[sessCfgPropName], {
                secure: propDefinition.secure
            });
        }

        await config.save(false);
        params.response.console.log(`Stored properties in ${config.layerActive().path}: ${profileProps.join(", ")}`);
        // Restore original active layer
        config.api.layers.activate(beforeLayer.user, beforeLayer.global);
    }

    /**
     * Finds the profile where auto-store properties should be saved.
     * @param params CLI handler parameters object
     * @param profileProps List of properties required in the profile schema
     * @returns Tuple containing profile type and name, or undefined if no
     *          profile was found
     */
    private static findActiveProfile(params: IHandlerParameters, profileProps: string[]): [string, string] | undefined {
        const profileTypes = [
            ...(params.definition.profile?.required || []),
            ...(params.definition.profile?.optional || [])
        ];

        for (const profType of profileTypes) {
            const profileMatch = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profType);
            if (profileMatch != null && profileProps.every(propName => propName in profileMatch.schema.properties)) {
                return [profType, ConfigUtils.getActiveProfileName(profType, params.arguments)];
            }
        }
    }

    /**
     * Retrieves token value that will be auto-stored into session config.
     * @param params CLI handler parameters object
     * @param sessCfg Session config with credentials for basic or cert auth
     * @param profilePath JSON path of profile containing tokenType
     * @returns True if auth handler was found and token was fetched
     */
    private static async fetchTokenForSessCfg(params: IHandlerParameters, sessCfg: { [key: string]: any }, profilePath: string): Promise<boolean> {
        const authHandlerClass = this.findAuthHandlerForProfile(profilePath, params.arguments);

        if (authHandlerClass == null) {
            return false;
        }

        const [promptParams, loginHandler] = authHandlerClass.getPromptParams();
        sessCfg.type = AUTH_TYPE_TOKEN;
        sessCfg.tokenType = promptParams.defaultTokenType;
        const baseSessCfg: ISession = { type: sessCfg.type };

        for (const propName of Object.keys(ImperativeConfig.instance.loadedConfig.baseProfile.schema.properties)) {
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            if (sessCfg[sessCfgPropName] != null) {
                (baseSessCfg as any)[sessCfgPropName] = sessCfg[sessCfgPropName];
            }
        }

        Logger.getAppLogger().info(`Fetching ${sessCfg.tokenType} for ${profilePath}`);
        sessCfg.tokenValue = await loginHandler(new Session(baseSessCfg));
        return true;
    }
}
