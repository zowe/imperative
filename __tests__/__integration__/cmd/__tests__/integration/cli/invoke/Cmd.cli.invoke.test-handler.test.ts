/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITestEnvironment } from "../../../../../../__src__/environment/doc/response/ITestEnvironment";
import { SetupTestEnvironment } from "../../../../../../__src__/environment/SetupTestEnvironment";
import { runCliScript } from "../../../../../../src/TestUtil";
import { ICommandResponse } from "../../../../../../../packages/cmd";
import * as fs from "fs";

// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;

describe("cmd-cli invoke test-handler", () => {
    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await SetupTestEnvironment.createTestEnv({
            cliHomeEnvVar: "CMD_CLI_CLI_HOME",
            testName: "cmd_cli_test_handler"
        });
    });

    it("should display the help", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_help.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    it("should fulfill the promise and complete the command successfully", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fulfill_promise.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    it("should fulfill the promise and complete the command successfully with a JSON response", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fulfill_promise_rfj.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");

        // Attempt to convert response to an object
        const respObj: ICommandResponse = JSON.parse(response.stdout.toString());

        // Check the fields of the response object
        expect(respObj.success).toBe(true);
        expect(respObj.stdout).toMatchSnapshot();
        expect(respObj.message).toBe("");
        expect(respObj.stderr).toBe("");
        expect(respObj.error).toBeUndefined();
    });

    it("should fail the command if fail-with-message is specified and no message is supplied", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_with_message_invalid.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stdout.toString()).toBe("");
        expect(response.stderr.toString()).toMatchSnapshot();
    });

    it("should fail the command if fail-with-message is specified with a message", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_with_message.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stdout.toString()).toBe("");
        expect(response.stderr.toString()).toMatchSnapshot();
    });

    it("should fail the command if fail-with-message is specified with a message and produce a JSON response", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_with_message_rfj.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);

        // Attempt to convert response to an object
        const respObj: ICommandResponse = JSON.parse(response.stdout.toString());

        // Check the fields of the response object
        expect(respObj.success).toBe(false);
        expect(respObj.stdout).toBe("");
        expect(respObj.stderr).toMatchSnapshot();
        expect(respObj.error.msg).toMatchSnapshot();
    });

    it("should fail a handler if a generic error is thrown", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_ge.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);

        // Ensure that the stderr field matches the expected regex - some non-deterministic data is contained within
        const regex = fs.readFileSync(__dirname + "/__regex__/test-handler/invoke_test_handler_fail_ge.regex").toString();
        expect(new RegExp(regex, "g").test(response.stderr.toString())).toBe(true);
    });

    it("should fail a handler if a generic error is thrown and produce a JSON response", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_ge_rfj.sh", TEST_ENVIRONMENT.workingDir);

        // Script status code should be 1 from imperative cli
        expect(response.status).toBe(1);

        // Verify that we can parse the stdout string into a JSON object
        const respObj: ICommandResponse = JSON.parse(response.stdout.toString());

        // Check the properties of the object for correctness
        expect(respObj.success).toBe(false);
        expect(respObj.message).toBe("Unexpected Command Error: Fail with Error");
        expect(respObj.error.msg).toBe("Fail with Error");
        expect(respObj.error.stack).toContain("Error: Fail with Error");
        expect(respObj.stdout).toBe("");

        // Ensure that the stderr field matches the expected regex - some non-deterministic data is contained within
        const regex = fs.readFileSync(__dirname + "/__regex__/test-handler/invoke_test_handler_fail_ge.regex").toString();
        expect(new RegExp(regex, "g").test(respObj.stderr.toString())).toBe(true);
    });

    it("should fail a handler if an imperative error is thrown", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_ie.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stderr.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    it("should fail a handler if an imperative error is thrown and produce a JSON response", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_fail_ie_rfj.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);

        // Verify that we can parse the stdout string into a JSON object
        const respObj: ICommandResponse = JSON.parse(response.stdout.toString());

        // Check the properties of the object for correctness
        expect(respObj.success).toBe(false);
        expect(respObj.message).toBe("Fail with Imperative Error");
        expect(respObj.error.msg).toBe("Fail with Imperative Error");
        expect(respObj.stderr).toMatchSnapshot();
        expect(respObj.stdout).toBe("");
    });

    it("should invoke the handler and print a message if no options are specified", async () => {
        const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    // TODO - Not fulfilling the promise causes the process to end, but the command processor does not get control
    // TODO - and therefore cannot produce the response format?
    // eslint-disable-next-line jest/no-commented-out-tests
    // it("should invoke the handler and print a message if no options are specified and produce a JSON response", async () => {
    //     const response = runCliScript(__dirname + "/__scripts__/test-handler/test_handler_rfj.sh", TEST_ENVIRONMENT.workingDir);
    //     console.log(response.stdout.toString());
    //     expect(response.status).toBe(0);
    //     expect(response.stderr.toString()).toBe("");

    //     // Verify that we can parse the stdout string into a JSON object
    //     const respObj: ICommandResponse = JSON.parse(response.stdout.toString());

    //     // Check the properties of the object for correctness
    //     expect(respObj.success).toBe(true);
    // });
});
