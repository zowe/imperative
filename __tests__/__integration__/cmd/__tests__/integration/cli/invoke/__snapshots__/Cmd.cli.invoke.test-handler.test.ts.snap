// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cmd-cli invoke test-handler should display the help 1`] = `
"========================TEST-HANDLER HELP================================

 COMMAND NAME
 ------------

   test-handler

 DESCRIPTION
 -----------

   Test handler that returns a promise. Tests rejecting and fulfilling the promise
   via the methods.

 USAGE
 -----

   cmd-cli invoke test-handler [options]

 OPTIONS
 -------

   --fail  (boolean)

      Fail the handler by invoking the reject method.

   --fail-with-message  (string)

      Fail the handler by invoking the reject method with a message.

   --fail-with-error  (boolean)

      Fail the handler by throwning a generic 'Error'.

   --fail-with-imperative-error  (boolean)

      Fail the handler by throwning an 'Imperative Error'.

   --fulfill-promise  (boolean)

      Fail the handler by throwning an 'Imperative Error'.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text

   --help-examples  (boolean)

      Display examples for all the commands in a group

   --help-web  | --hw (boolean)

      Display HTML help in browser

========================TEST-HANDLER HELP RFJ============================
{
  \\"success\\": true,
  \\"exitCode\\": 0,
  \\"message\\": \\"The help was constructed for command: test-handler.\\",
  \\"stdout\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   test-handler\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Test handler that returns a promise. Tests rejecting and fulfilling the promise\\\\n   via the methods.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   cmd-cli invoke test-handler [options]\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --fail  (boolean)\\\\n\\\\n      Fail the handler by invoking the reject method.\\\\n\\\\n   --fail-with-message  (string)\\\\n\\\\n      Fail the handler by invoking the reject method with a message.\\\\n\\\\n   --fail-with-error  (boolean)\\\\n\\\\n      Fail the handler by throwning a generic 'Error'.\\\\n\\\\n   --fail-with-imperative-error  (boolean)\\\\n\\\\n      Fail the handler by throwning an 'Imperative Error'.\\\\n\\\\n   --fulfill-promise  (boolean)\\\\n\\\\n      Fail the handler by throwning an 'Imperative Error'.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n   --help-examples  (boolean)\\\\n\\\\n      Display examples for all the commands in a group\\\\n\\\\n   --help-web  | --hw (boolean)\\\\n\\\\n      Display HTML help in browser\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   test-handler\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   Test handler that returns a promise. Tests rejecting and fulfilling the promise\\\\n   via the methods.\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   cmd-cli invoke test-handler [options]\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --fail  (boolean)\\\\n\\\\n      Fail the handler by invoking the reject method.\\\\n\\\\n   --fail-with-message  (string)\\\\n\\\\n      Fail the handler by invoking the reject method with a message.\\\\n\\\\n   --fail-with-error  (boolean)\\\\n\\\\n      Fail the handler by throwning a generic 'Error'.\\\\n\\\\n   --fail-with-imperative-error  (boolean)\\\\n\\\\n      Fail the handler by throwning an 'Imperative Error'.\\\\n\\\\n   --fulfill-promise  (boolean)\\\\n\\\\n      Fail the handler by throwning an 'Imperative Error'.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n   --help-examples  (boolean)\\\\n\\\\n      Display examples for all the commands in a group\\\\n\\\\n   --help-web  | --hw (boolean)\\\\n\\\\n      Display HTML help in browser\\\\n\\\\n\\"
}"
`;

exports[`cmd-cli invoke test-handler should fail a handler if an imperative error is thrown 1`] = `
"Command Error:
Fail with Imperative Error
"
`;

exports[`cmd-cli invoke test-handler should fail a handler if an imperative error is thrown 2`] = `""`;

exports[`cmd-cli invoke test-handler should fail a handler if an imperative error is thrown and produce a JSON response 1`] = `
"Command Error:
Fail with Imperative Error
"
`;

exports[`cmd-cli invoke test-handler should fail the command if fail-with-message is specified and no message is supplied 1`] = `
"
Syntax Error:
No value specified for option:
--fail-with-message

This option requires a value of type:
string

Option Description:
Fail the handler by invoking the reject method with a message.

Use \\"cmd-cli invoke test-handler --help\\" to view command description, usage, and options.
"
`;

exports[`cmd-cli invoke test-handler should fail the command if fail-with-message is specified with a message 1`] = `
"Command Error:
Failure message from option!
"
`;

exports[`cmd-cli invoke test-handler should fail the command if fail-with-message is specified with a message and produce a JSON response 1`] = `
"Command Error:
Failure message from option!
"
`;

exports[`cmd-cli invoke test-handler should fail the command if fail-with-message is specified with a message and produce a JSON response 2`] = `"Failure message from option!"`;

exports[`cmd-cli invoke test-handler should fulfill the promise and complete the command successfully 1`] = `
"Fulfilling the promise...
"
`;

exports[`cmd-cli invoke test-handler should fulfill the promise and complete the command successfully with a JSON response 1`] = `
"Fulfilling the promise...
"
`;

exports[`cmd-cli invoke test-handler should invoke the handler and print a message if no options are specified 1`] = `
"No options specified, will not invoke anything!
"
`;
